### PR TITLE
Slamd 120

### PR DESCRIPTION
**Pull request by StefanLueders**
Added support for partially labelled data. Originally I planned to get rid of label_index and nolabel_index, but it turned to be smarter to keep them and add indices for partially labelled data and predicted rows (rows with at least one NA value).

When reviewing this, please do some manual tests to ensure that partially labelled data produces sensible results. I think it does, but a second opinion would be great.

### CODE REVIEW

File: slamd/discovery/processing/experiment/experiment_conductor.py [hunk 1]
```diff
@@ -40,25 +41,34 @@ def _fit_model_and_predict(cls, exp):
...
+            rows_to_predict = exp.features_df.loc[index_unlabelled].values
             prediction, uncertainty = regressor.predict(rows_to_predict, return_std=True)
 
-            predictions[target] = prediction
-            uncertainties[target] = uncertainty
+            predictions.loc[index_unlabelled, target] = prediction
+            uncertainties.loc[index_unlabelled, target] = uncertainty
+
+            # Determine rows for which the current target is labelled, but others aren't
+            index_only_curr_labelled = exp.index_partially_labelled.intersection(index_labelled)
```
Comment (RugTehseen) [inline slamd/discovery/processing/experiment/experiment_conductor.py:64]: are there any edge cases you could think of where the intersection is empty. If yes, do we need to include exception handling here?

File: tests/discovery/processing/test_discovery_service.py [hunk 2]
```diff
@@ -212,6 +212,31 @@ def mock_save_tsne_plot_data(tsne_plot_data):
     assert mock_save_tsne_plot_data_called_with == 'Saving TSNE Plot Data'
 
 
+def test_run_experiment_with_partially_labelled_data(monkeypatch):
```
Comment (RugTehseen) [inline tests/discovery/processing/test_discovery_service.py:215]: It is not directly a part of this story, but since we are adding more tests for the run_experiment part of the discovery_service, we could think about creating a dedicated testfile, e.g. test_discovery_service_experiment.py and move all tests related to the experiment to that file. This way the test files are a little more comprehensible.

**Review by RugTehseen — COMMENTED** [review]
Added some comments about testing and exception handling. Aspects related to the functionality itself were discussed and look good for now.

File: slamd/discovery/processing/experiment/experiment_conductor.py (hunk 1 above)
Comment (StefanLueders) [reply slamd/discovery/processing/experiment/experiment_conductor.py:64]: The exception being empty is the standard case, since it is only non-empty if we have partial labels. Afaik pandas is generally smart enough to work with empty indices; the fact that our tests work shows that we don't need exception handling here.

File: slamd/discovery/processing/experiment/experiment_data.py [hunk 3]
```diff
@@ -45,10 +45,17 @@ def apriori_df(self):
...
 
     @property
-    def nolabel_index(self):
+    def index_none_labelled(self):
         return self.dataframe.index[self.targets_df.isnull().all(axis=1)]
 
     @property
-    def label_index(self):
+    def index_partially_labelled(self):
+        return self.dataframe.index.difference(self.index_all_labelled).difference(self.index_none_labelled)
```
[the remark this answers is not in the document]
Comment (StefanLueders) [reply slamd/discovery/processing/experiment/experiment_data.py:53]: Added unit tests for indices, I don't think they are necessary for the dataframes

[the remark this answers is not in the document]
Comment (RugTehseen) [reply slamd/discovery/processing/experiment/experiment_data.py:53]: the new tests look good

File: tests/discovery/processing/test_discovery_service.py (hunk 2 above)
Comment (StefanLueders) [reply tests/discovery/processing/test_discovery_service.py]: Done

File: tests/discovery/processing/test_discovery_service_run_experiment.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+def test_run_experiment_with_partially_labelled_data(monkeypatch):
+    mock_save_prediction_called_with = None
+
+    def mock_save_prediction(prediction):
+        nonlocal mock_save_prediction_called_with
+        mock_save_prediction_called_with = prediction
+        return None
+
+    mock_save_tsne_plot_data_called_with = None
```
Comment (RugTehseen) [inline tests/discovery/processing/test_discovery_service_run_experiment.py:120]: the mock is unused (no assertions) here. Either you can also insert on it or remove the nonlocal var and simply return None in the corresponding mock's method. Both is fine for me as we are alreading testing this mock in other tests.

**Review by RugTehseen — COMMENTED** [review]
There is an unused nonlocal mock var which should either be deleted or asserted on.

File: tests/discovery/processing/test_discovery_service_run_experiment.py (hunk 4 above)
Comment (RugTehseen) [reply tests/discovery/processing/test_discovery_service_run_experiment.py]: removed the unused mock

**Review by RugTehseen — APPROVED** [review]
looks good!
